### PR TITLE
Tune performance of optimize_1q_decomposition

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

This commit tunes the performance of the optimize_1q_decomposition pass.
Previously, the bottleneck of the pass was in creating the operator to
decompose because we were creating an unecessary circuit and then having
the Operator class convert that to a matrix which involved running
to_instruction() and lots of copying. Instead this commit switches the
pass to just go straight to an array or operator (if __array__ isn't
defined) for each gate and compose that directly onto the operator.
Doing this ends up being faster than using a circuit as an
intermediate class.

### CODE REVIEW

**Review by chriseclectic — CHANGES_REQUESTED** [review]
I think it would be better to move the if checking into the init method of Operator:
```python
elif isinstance(data, (QuantumCircuit, Instruction)):
    if hasattr(data, '__array__'):
        self._data = np.array(data, dtype=complex)
    else:
        self._data = self._init_instruction(data).data
```
And then the code here can be written as:
```python
operator = Operator(run[0].op)
for gate in run[1:]:
    operator = operator.compose(gate.op)
```

**Comment by chriseclectic** [comment]
Actually, rather than what I wrote above it would probably be even easier to update the operator to only modify `_init_instruction` to be
```python
@classmethod
    def _init_instruction(cls, instruction):
        """Convert a QuantumCircuit or Instruction to an Operator."""
        if hasattr(instruction, '__array__'):
            return Operator(np.array(instruction, dtype=complex))
       
       ... original code of this method ...
```

**Review by chriseclectic — APPROVED** [review]
LGTM.

Only possible issue is I'm not sure if `run` can be an empty list and we have to check for that edge case before calling `run[0]`?

**Comment by mtreinish** [comment]
> LGTM.
> 
> Only possible issue is I'm not sure if `run` can be an empty list and we have to check for that edge case before calling `run[0]`?

`run` should never end up being empty. `collect_1q_runs()` just calls retworkx's `collect_runs()`
 which only adds a run list to the output if it's not empty: https://github.com/Qiskit/retworkx/blob/master/src/lib.rs#L1166-L1168

File: qiskit/dagcircuit/dagcircuit.py [hunk 1]
```diff
@@ -826,10 +827,15 @@ def _full_pred_succ_maps(self, pred_map, succ_map, input_circuit,
         return full_pred_map, full_succ_map
 
     def __eq__(self, other):
-        if (
-                self.global_phase != other.global_phase
-                or self.calibrations != other.calibrations
-        ):
+        try:
```
Comment (kdk) [inline qiskit/dagcircuit/dagcircuit.py:830]: Can you add a comment explaining the need for the `try`?

File: qiskit/transpiler/passes/optimization/optimize_1q_decomposition.py [hunk 2]
```diff
@@ -69,16 +68,14 @@ def run(self, dag):
...
-            operator = Operator(qc)
+            operator = Operator(run[0].op)
+            for gate in run[1:]:
+                operator = operator.compose(gate.op)
             for decomposer in self.basis:
                 new_circs.append(decomposer(operator))
             if new_circs:
                 new_circ = min(new_circs, key=lambda circ: circ.depth())
-                if qc.depth() > new_circ.depth():
+                if len(run) > new_circ.depth():
```
Comment (kdk) [inline qiskit/transpiler/passes/optimization/optimize_1q_decomposition.py:78]:
```suggestion
                if len(run) > len(new_circ):
```
and similarly in the line above?

Comment (mtreinish) [reply qiskit/transpiler/passes/optimization/optimize_1q_decomposition.py:78]: Yeah, in both places makes sense

Comment (mtreinish) [reply qiskit/transpiler/passes/optimization/optimize_1q_decomposition.py]: Done in: https://github.com/Qiskit/qiskit-terra/pull/5682/commits/7b6965a4c89c71557370d02e78439c8a0d37abc0

File: qiskit/dagcircuit/dagcircuit.py (hunk 1 above)
Comment (mtreinish) [reply qiskit/dagcircuit/dagcircuit.py:833]: Done in: https://github.com/Qiskit/qiskit-terra/pull/5682/commits/0d2187498573c5840e1184cf9fe7f2429af13dc6

**Comment by mtreinish** [comment]
I removed the backport potential tag on this, it's a bit too risky imo from a regression perspective for a backport especially when weight against the fact that it's just a performance improvement.

**Review by kdk — APPROVED** [review]

**Pull request merged**